resqpy/grid/grid_functions.py

```diff
@@ -0,0 +1,187 @@
+
+import resqpy.olio.xml_et as rqet
+import resqpy.property as rprop
+import resqpy
+import numpy as np
+import resqpy.olio.point_inclusion as pip
+
+# 'private' function
+
+
+def _add_to_kelp_list(extent_kji, kelp_list, face_axis, ji):
+    if isinstance(face_axis, bool):
+        face_axis = 'J' if face_axis else 'I'
+    # ignore external faces
+    if face_axis == 'J':
+        if ji[0] < 0 or ji[0] >= extent_kji[1] - 1:
+            return
+    elif face_axis == 'I':
+        if ji[1] < 0 or ji[1] >= extent_kji[2] - 1:
+            return
+    else:  # ji is actually kj or ki
+        assert face_axis == 'K'
+        if ji[0] < 0 or ji[0] >= extent_kji[0] - 1:
+            return
+    pair = ji
+    if pair in kelp_list:
+        return  # avoid duplication
+    kelp_list.append(pair)
+
+
+def establish_zone_property_kind(model):
+    """Returns zone local property kind object, creating the xml and adding as part if not found in model."""
+
+    zone_pk_uuid = model.uuid(obj_type = 'LocalPropertyKind', title = 'zone')
+    if zone_pk_uuid is None:
+        zone_pk = rprop.PropertyKind(model, title = 'zone', parent_property_kind = 'discrete')
+        zone_pk.create_xml()
+    else:
+        zone_pk = rprop.PropertyKind(model, uuid = zone_pk_uuid)
+    return zone_pk
+
+
+def extent_kji_from_root(root_node):
+    """Returns kji extent as stored in xml."""
+
+    return (rqet.find_tag_int(root_node, 'Nk'), rqet.find_tag_int(root_node, 'Nj'), rqet.find_tag_int(root_node, 'Ni'))
+
+
+def grid_flavour(grid_root):
+    """Returns a string indicating type of grid geometry, currently 'IjkGrid' or 'IjkBlockGrid'."""
+
+    if grid_root is None:
+        return None
+    em = rqet.load_metadata_from_xml(grid_root)
+    flavour = em.get('grid_flavour')
+    if flavour is None:
+        node_type = rqet.node_type(grid_root, strip_obj = True)
+        if node_type == 'IjkGridRepresentation':
+            if rqet.find_tag(grid_root, 'Geometry') is not None:
+                flavour = 'IjkGrid'
+            else:
+                flavour = 'IjkBlockGrid'  # this might cause issues
+        elif node_type == 'UnstructuredGridRepresentation':
+            cell_shape = rqet.find_nested_tags_text(grid_root, ['Geometry', 'CellShape'])
+            if cell_shape is None or cell_shape == 'polyhedral':
+                flavour = 'UnstructuredGrid'
+            elif cell_shape == 'tetrahedral':
+                flavour = 'TetraGrid'
+            elif cell_shape == 'hexahedral':
+                flavour = 'HexaGrid'
+            elif cell_shape == 'pyramidal':
+                flavour = 'PyramidGrid'
+            elif cell_shape == 'prism':
+                flavour = 'PrismGrid'
+    return flavour
+
+
+def is_regular_grid(grid_root):
+    """Returns True if the xml root node is for a RegularGrid."""
+
+    return grid_flavour(grid_root) == 'IjkBlockGrid'
+
+
+def any_grid(parent_model, grid_root = None, uuid = None, find_properties = True):
```

Comment:
I feel this function, together with grid_flavour() and is_regular_grid() should go in a separate module, as they apply to several classes of grid, whereas other functions are specific to Grid.